File: classes/@ML_model/ML_model.m
classdef ML_model < neuron_model
    
    % ML_model   Neuron represented by the Morris-Lecar model
    %
    % The Neuron is modeld as follow:
    %   _
    %  |
    %  | dV/dt = (-gl*(V - Vl) - gCa*Minf*(V - VCa) - gK*N*(V - Vk ) + I - I_{ext})/CM
    % <
    %  | dN/dt = lamda_N*(Ninf-N)
    %  |_-
    %
    %   Minf = 0.5*(1 + tanh((V - V1 )/V2 ));
    %   Ninf = 0.5*(1 + tanh((V - V3 )/V4 ));
    %   lamda_N = phi*cosh((V - V3 )/(2*V4) );
    %
    %   CM = 5;
    %   gK = 8;
    %   gl = 2;
    %   VCa = 120;
    %   Vk = -80;
    %   Vl = -60;
    %   V1 = -1.2;
    %   V2 = 18;
    %
    % Ther model has two state variable; V represent the membrane potential.
    % gCa,V3,V4,phi and I are parameters that allows to change neuron behaviours.
    %
    % OBJ = ML_model()
    % Builds an ML_model object OBJ with all parameters equal to 0.
    %
    % OBJ = ML_model(CM,gCa,V3,V4,phi,gl,Vl, I)
    % Builds an ML_model object OBJ with user assigned parameters value
    %
    %
    %   ML_model methods:
    %   getXdot - omputes the derivative of the state
    %   sim - simulates the neuron.
    %   simplot - simulates the neuron system and plots time evolution of
    %             states.
    %   disp - displays some information about the FN_relaxation_model object
    %   get_gCa - gets the value of parameter gCa
    %   get_V3 - gets the value of parameter V3
    %   get_V4 - gets the value of parameter V4
    %   get_I - gets the value of parameter I
    %   set_gCa - sets the value of parameter gCa
    %   set_V3 - sets the value of parameter V3
    %   set_V4 - sets the value of parameter V4
    %   set_I - sets the value of parameter I
    %   getJacobian -Computes the Jacobian of the vector field in point x
    %   generateC - Generates C files for the computation of model vector field
    %   getJacobian -Computes the Jacobian of the vector field in point x
    %   generateC - Generates C files for the computation of model vector field
    %
    % The ML_model object is derived from neuron_model and
    % inherits all its methods.
    %
    % See also FN_relaxation_model, neuron_model
    %
    % Contributors:
    %
    % Matteo Lodi (matteo.lodi@edu.unige.it)
    %
    % Copyright (C) 2016 University of Genoa, Italy.
    
    % Legal note:
    % This program is free software; you can redistribute it and/or
    % modify it under the terms of the GNU General Public
    % License as published by the Free Software Foundation; either
    % version 2.1 of the License, or (at your option) any later version.
    %
    % This program is distributed in the hope that it will be useful,
    % but WITHOUT ANY WARRANTY; without even the implied warranty of
    % MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    % General Public License for more details.
    %
    % You should have received a copy of the GNU General Public
    % License along with this library; if not, write to the
    % Free Software Foundation, Inc.,
    % 59 Temple Place, Suite 330,
    % Boston, MA  02111-1307  USA
    
    %properties
    properties(Access = protected)
        gCa = 0;
        V3 = 0;
        V4 = 0;
        phi = 0;
        I = 0;
        CM = 0;
        gL = 0;
        EL = 0;
    end
    
    methods
        function obj = ML_model(varargin)
            obj.nx = 2;
            obj.xnames = {'V','N'};
            obj.modelName = 'Morris-Lecar';
            obj.isContinuous = true;
            if nargin == 8
                obj.CM = varargin{1};
                obj.gCa = varargin{2};
                obj.V3 = varargin{3};
                obj.V4 = varargin{4};
                obj.phi = varargin{5};
                obj.gL = varargin{6};
                obj.EL = varargin{7};
                obj.I = varargin{8};
            end
        end
        
        % Other methods
        x_dot = getXdot(object,t,x,varargin);
        disp(object);
        str = getCbuilder(object);
        
        % get methods
        CM = get_CM(object);
        gCa = get_gCa(object);
        V3 = get_V3(object);
        V4 = get_V4(object);
        phi = get_phi(object);
        gL = get_gL(object);
        EL = get_EL(object);
        I = get_I(object);
        
        % set methods
        object = set_CM(object,CM);
        object = set_gCa(object,gCa);
        object = set_V3(object,V3);
        object = set_V4(object,V4);
        object = set_phi(object,phi);
        object = set_gL(object,gL);
        object = set_EL(object,EL);
        object = set_I(object,I);
    end
    
end

File: c_file/src/ML_model.cpp
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

/* 
 * File:   ML_model.cpp
 * Author: teo
 * 
 * Created on July 12, 2017, 10:21 AM
 */

#include "../inc/ML_model.h"

ML_model::ML_model() 
{
    this->nx = 2;
    this->CM = 0;
    this->gCa = 0;
    this->V3 = 0;
    this->V4 = 0;
    this->phi = 0;
    this->gl = 0;
    this->Vl = 0;
    this->I = 0;
}

ML_model::ML_model(double CM,double gCa,double V3,double V4,double phi,double gl,double Vl,double I)
{
    this->nx = 2;
    this->CM = CM;
    this->gCa = gCa;
    this->V3 = V3;
    this->V4 = V4;
    this->phi = phi;
    this->gl = gl;
    this->Vl = Vl;
    this->I = I;
}

void ML_model::getXdot(double t, double *x, double *xdot,double *Iext)
{
double gCa = this->gCa;
double I = this->I;
double V3 = this->V3;
 double V4 = this->V4;
double phi = this->phi;
double Vl = this->Vl;
double gl = this->gl;

double CM = this->CM;

const double gK = 8;;
const double VCa = 120;
const double Vk = -80;
const double V1 = -1.2;
const double V2 = 18;

double Minf = 0.5*(1 + tanh((x[0] - V1 )/V2 ));
double Ninf = 0.5*(1 + tanh((x[0] - V3 )/V4 ));
double lamda_N = phi*cosh((x[0] - V3 )/(2*V4) );

xdot[1] = lamda_N*(Ninf-x[1]);
xdot[0] = (-gl*(x[0] - Vl) - gCa*Minf*(x[0] - VCa) - gK*x[1]*(x[0] - Vk ) + I + Iext[0])/CM;

}

ML_model::~ML_model() 
{
}

